### PR TITLE
8280123: C2: Infinite loop in CMoveINode::Ideal during IGVN

### PR DESCRIPTION
**Pull request by iwanowww**
There's a discrepancy between `CMoveNode::Ideal()` and `CMoveINode::Ideal()` in the way how constants are detected:
the former looks for `Con` nodes (`in(IfFalse)->is_Con()`) while the latter checks node types (e.g., `phase->type(in(IfFalse)) == TypeInt::ZERO`). It leads to infinite loop during IGVN (where  `CMoveNode::Ideal()` and `CMoveINode::Ideal()` repeatedly rotate `CMoveI` node) because `CMoveNode::Ideal()` don't consider `CastII` to a constant as a constant while `CMoveINode::Ideal()` does.

The fix migrates `CMoveNode::Ideal()` away from `Node::is_Con()` to node types.

Testing: hs-tier1 - hs-tier4

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280123](https://bugs.openjdk.java.net/browse/JDK-8280123): C2: Infinite loop in CMoveINode::Ideal during IGVN

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to d29bbdd159cb80c267be182e68866c64239ff2e9
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to d29bbdd159cb80c267be182e68866c64239ff2e9
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to d29bbdd159cb80c267be182e68866c64239ff2e9

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7133/head:pull/7133` \
`$ git checkout pull/7133`

Update a local copy of the PR: \
`$ git checkout pull/7133` \
`$ git pull https://git.openjdk.java.net/jdk pull/7133/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7133`

View PR using the GUI difftool: \
`$ git pr show -t 7133`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7133.diff">https://git.openjdk.java.net/jdk/pull/7133.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Looks good.

File: src/hotspot/share/opto/movenode.cpp [hunk 1]
```diff
@@ -75,17 +75,24 @@
...
   assert(in(Condition) != this &&
-         in(IfFalse) != this &&
-         in(IfTrue) != this, "dead loop in CMoveNode::Ideal" );
-  if( phase->type(in(Condition)) == Type::TOP )
-  return NULL; // return NULL when Condition is dead
-
-  if( in(IfFalse)->is_Con() && !in(IfTrue)->is_Con() ) {
-    if( in(Condition)->is_Bool() ) {
+         in(IfFalse)   != this &&
+         in(IfTrue)    != this, "dead loop in CMoveNode::Ideal" );
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/movenode.cpp:87]:
```suggestion
         in(IfTrue)    != this, "dead loop in CMoveNode::Ideal");
```

File: src/hotspot/share/opto/movenode.cpp [hunk 2]
```diff
@@ -75,17 +75,24 @@
...
+  if (phase->type(in(Condition)) == Type::TOP ||
+      phase->type(in(IfFalse))   == Type::TOP ||
+      phase->type(in(IfTrue))    == Type::TOP) {
+    return NULL;
+  }
+  // Canonicalize the node by moving constants to the right input.
+  if (phase->type(in(IfFalse))->singleton() && !phase->type(in(IfTrue))->singleton()) {
+    if (in(Condition)->is_Bool()) {
       BoolNode* b  = in(Condition)->as_Bool();
       BoolNode* b2 = b->negate(phase);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/movenode.cpp:97]:
```suggestion
      BoolNode* b  = in(Condition)->as_Bool()->negate(phase);
```

File: src/hotspot/share/opto/movenode.cpp [hunk 3]
```diff
@@ -75,17 +75,24 @@
...
       return make(in(Control), phase->transform(b2), in(IfTrue), in(IfFalse), _type);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/movenode.cpp:98]:
```suggestion
      return make(in(Control), phase->transform(b), in(IfTrue), in(IfFalse), _type);
```

File: test/hotspot/jtreg/compiler/c2/TestCMoveInfiniteGVN.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/TestCMoveInfiniteGVN.java:25]:
```suggestion
 * @test
 * @key stress randomness
```

**Review by TobiHartmann — APPROVED** [review]
Looks good. I added some minor cleanup suggestions.

File: src/hotspot/share/opto/movenode.cpp [hunk 5]
```diff
@@ -75,17 +75,24 @@
...
+         in(IfFalse)   != this &&
+         in(IfTrue)    != this, "dead loop in CMoveNode::Ideal" );
+  if (phase->type(in(Condition)) == Type::TOP ||
+      phase->type(in(IfFalse))   == Type::TOP ||
+      phase->type(in(IfTrue))    == Type::TOP) {
+    return NULL;
+  }
+  // Canonicalize the node by moving constants to the right input.
+  if (phase->type(in(IfFalse))->singleton() && !phase->type(in(IfTrue))->singleton()) {
+    if (in(Condition)->is_Bool()) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/movenode.cpp:95]: Could be merged into a single `if`.

**Review by chhagedorn — APPROVED** [review]
Looks good!

**Comment by iwanowww** [comment]
Thanks for the reviews, Vladimir, Tobias, and Christian.

/integrate

**Pull request closed without merging**